iotfunctions/data_quality.py

```diff
@@ -0,0 +1,183 @@
+
+        recommended values
+        m (2, 3)
+        r (0.1, 2.5) * standard deviation
+
+        :returns float
+        """
+        def sampen(L, m, r):
+            N = len(L)
+
+            # Split time series and save all templates of length m
+            xmi = np.array([L[i: i + m] for i in range(N - m)])
+            xmj = np.array([L[i: i + m] for i in range(N - m + 1)])
+
+            # Save all matches minus the self-match, compute B
+            B = np.sum([np.sum(np.abs(xmii - xmj).max(axis=1) <= r) - 1 for xmii in xmi])
+
+            # Similar for computing A
+            m += 1
+            xm = np.array([L[i: i + m] for i in range(N - m + 1)])
+
+            A = np.sum([np.sum(np.abs(xmi - xm).max(axis=1) <= r) - 1 for xmi in xm])
+
+            # Return SampEn
+            return -np.log(A / B)
+
+        return sampen(series.to_list(), m=2, r=0.2 * series.std())
+
+    @staticmethod
+    def stationarity(series):
+        """
+        A time series is Stationary when it's mean, variance, co-variance do not change over time.
+        Time-invariant process are requirements of statistical models for forecasting problems
+        Can indicate spurious causation between variable dependent on time
+
+        Reference:
+        https://www.statsmodels.org/stable/examples/notebooks/generated/stationarity_detrending_adf_kpss.html
+        performs adf and kpss stationarity tests
+        :returns str (Not Stationary, Stationary, Trend Stationary, Difference Stationary, Constant Data)
+        """
+        stationary_type = {
+            # adf stationary, kpss stationary
+            (False, False): 'Not Stationary',
+            (False, True): 'Trend Stationary',
+            (True, False): 'Difference Stationary',
+            (True, True): 'Stationary',
+            (np.nan, np.nan): 'Constant Data',
+            'NoCompute': 'Not Enough Data for stationarity test'
+        }
+        if len(series) < 4:
+            return stationary_type['NoCompute']
+
+        significance_level = 0.05  # p > 5% fail to reject the null hypothesis
+        # adf test; H0: series has unit root (non-stationary)
+        adf_statistic, adf_p_value, _, _, _, _ = adfuller(series)
+        adf_stationary = np.nan
+        if not math.isnan(adf_p_value):
+            adf_stationary = bool(adf_p_value < significance_level)  # reject null
+
+        # kpss test; H0: process is trend stationary
+        kpss_statistic, kpss_p_value, _, _ = kpss(series)
+        kpss_stationary = np.nan
+        if not math.isnan(kpss_p_value):
+            kpss_stationary = bool(kpss_p_value >= significance_level)  # fail to reject null
+
+        return stationary_type[adf_stationary, kpss_stationary]
+
+    @staticmethod
+    def stuck_at_zero(series):
+        """
+        A time series signal stuck at zero contains no information
+
+        :returns bool
+        """
+        return bool(np.count_nonzero(series.to_numpy()) < 1)
```

Comment:
How about using np.all_close to account for values that are almost zero (up to an error of a 10^-8 or so)